internal/cmd/launcher.go

```diff
@@ -0,0 +1,348 @@
+package cmd
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"os"
+	"os/exec"
+	"strings"
+
+	"github.com/Masterminds/semver/v3"
+	"github.com/grafana/k6deps"
+	"github.com/grafana/k6provider"
+	"go.k6.io/k6/cloudapi"
+	"go.k6.io/k6/cmd/state"
+	"go.k6.io/k6/internal/build"
+)
+
+var (
+	errScriptNotFound     = errors.New("script not found")
+	errUnsupportedFeature = errors.New("not supported")
+)
+
+// commandExecutor executes the requested k6 command line command.
+// It abstract the execution path from the concrete binary.
+type commandExecutor interface {
+	run(*state.GlobalState)
+}
+
+// Launcher is a k6 launcher. It analyses the requirements of a k6 execution,
+// then if required, it provisions a binary executor to satisfy the requirements.
+type Launcher struct {
+	// gs is the global state of k6.
+	gs *state.GlobalState
+
+	// provision generates a custom binary from the received list of dependencies
+	// with their constrains, and it returns an executor that satisfies them.
+	provision func(*state.GlobalState, k6deps.Dependencies) (commandExecutor, error)
+
+	// commandExecutor executes the requested k6 command line command
+	commandExecutor commandExecutor
+}
+
+// NewLauncher creates a new Launcher from a GlobalState using the default fallback and provision functions
+func NewLauncher(gs *state.GlobalState) *Launcher {
+	defaultExecutor := &currentBinary{}
+	return &Launcher{
+		gs:              gs,
+		provision:       k6buildProvision,
+		commandExecutor: defaultExecutor,
+	}
+}
+
+// Launch executes k6 either by launching a provisioned binary or defaulting to the
+// current binary if this is not necessary.
+// If the fhe fallback is called, it can exit the process so don't assume it will return
```

Comment:
```suggestion
// If the fallback is called, it can exit the process so don't assume it will return
```

Reply:
There's no fallback anymore. I fixed the command to make reference to the `commandExecutor`